project/src/main/java/com/google/sps/data/PlacesFetcher.java

```diff
@@ -104,25 +107,40 @@ public PlacesFetcher(GeoApiContext geoApiContext) {
      *     for places or for places details
      */
     public ImmutableList<Place> fetch(UserPreferences preferences) throws FetcherException {
-        PlacesSearchResult[] placesSearchResult;
+        Map<String, Set<String>> placesSearchResults = new HashMap<>();
+        PlacesSearchResult[] resultsForCuisine;
+        // If user didn't choose any cuisines, search on all possible cuisines
+        Collection<String> cuisines =
+            (preferences.cuisines().isEmpty())
+                ? CUISINE_TO_SEARCH_WORDS.keySet() : preferences.cuisines();
         int attemptsCounter = 0;
         do {
-            attemptsCounter++;
-            try {
-                placesSearchResult = getPlacesSearchResults(
-                    genTextSearchRequest(preferences, INIT_SEARCH_RADIUS_M * attemptsCounter));
-            } catch (ApiException | InterruptedException | IOException | IllegalStateException e) {
-                throw new FetcherException("Couldn't fetch places from Places API", e);
+            for (String cuisine: cuisines) {
+                attemptsCounter++;
+                try {
+                    resultsForCuisine = getPlacesSearchResults(genTextSearchRequest(
+                        preferences, INIT_SEARCH_RADIUS_M * attemptsCounter, cuisine));
+                } catch (ApiException
+                        | InterruptedException
+                        | IOException
+                        | IllegalStateException e) {
+                    throw new FetcherException("Couldn't fetch places from Places API", e);
+                }
+                for (PlacesSearchResult result : resultsForCuisine) {
+                    placesSearchResults.computeIfAbsent(
+                        result.placeId, k -> new HashSet<>()).add(cuisine);
+                }
```

Comment:
It looks really good!
I have a point I'm not sure about - from what I understand about computeIfAbsent, if the key already exists, it does nothing. So doesn't that mean a single place can only be added to the HashMap with one cuisine?

Reply:
No, that's exactly why I chose to use this - so a single place can have multiple cuisines :)
It's working and tested.
The computeIfAbsent creates a new Set of cuisines if the result wasn't already found, but if the results already exists then its cuisines set is returned and the new cuisine is added to it.